lib/nrf_modem_lib/nrf_modem_lib.c

```diff
@@ -151,12 +154,23 @@ int nrf_modem_lib_bootloader_init(void)
 
 int nrf_modem_lib_shutdown(void)
 {
+	int ret;
+	uint16_t mode;
+
 	LOG_DBG("Shutting down modem library");
 	STRUCT_SECTION_FOREACH(nrf_modem_lib_shutdown_cb, e) {
 		LOG_DBG("Modem shutdown callback: %p", e->callback);
 		e->callback(e->context);
 	}
 
+	/* The application must set CFUN=0 before calling nrf_modem_shutdown. */
+	ret = nrf_modem_at_scanf(AT_CFUN_READ, "+CFUN: %hu", &mode);
+	if (ret == 1 && mode != AT_CFUN0_VAL) {
+		LOG_WRN("Application should set minimal functional mode (CFUN=0) before
```

Comment:
Do we really need this warning?
This function already check CFUN, why should we need to duplicate this in the application?

Reply:
Perhaps not. The reasoning for it is that the application sets the functional mode to CFUN=1 etc. so it should be responsible for power off (CFUN=0) as well. But we add it here to prevent issues if the application has not done it already. One alternative we looked at was to return an error, but we want the application to be able to use the NRF_MODEM_LIB_SHUTDOWN hook for doing CFUN=0.

Reply:
@lemrey Any thoughts?

Reply:
We print a warning because the application should be responsible of controlling the modem's functional mode. We don't want the application to rely on `nrf_modem_lib_shutdown` to set the functional mode. Before shutting down the modem, the application should CFUN to zero; if it doesn't it's a bug in the application code.